airflow/api_fastapi/core_api/routes/ui/dags.py

```diff
@@ -0,0 +1,150 @@
+    only_active: QueryOnlyActiveFilter,
+    paused: QueryPausedFilter,
+    last_dag_run_state: QueryLastDagRunStateFilter,
+    dag_runs_limit: QueryLimit,
+    session: Annotated[Session, Depends(get_session)],
+) -> RecentDAGRunsCollectionResponse:
+    """Get recent DAG runs."""
+    recent_runs_subquery = (
+        select(
+            DagRun.dag_id,
+            DagRun.execution_date,
+            func.rank()
+            .over(
+                partition_by=DagRun.dag_id,
+                order_by=DagRun.execution_date.desc(),
+            )
+            .label("rank"),
+        )
+        .order_by(DagRun.execution_date.desc())
+        .subquery()
+    )
+    dags_with_recent_dag_runs_select = (
+        select(
+            DagModel.dag_id,
+            # should select DagModel fields that `get_dags` in public endpoint will filter at
+            DagModel.dag_display_name,
+            DagModel.next_dagrun,
+            recent_runs_subquery.c.execution_date,
+            DagRun.start_date,
+            DagRun.end_date,
+            DagRun.state,
+            DagRun.data_interval_start,
+            DagRun.data_interval_end,
+        )
+        .join(DagModel, DagModel.dag_id == recent_runs_subquery.c.dag_id)
+        # `last_run_state` and `last_run_start_date` query params of`get_dags` endpoint
+        # needs `DagRun` to order_by
+        .join(
+            DagRun,
+            and_(
+                DagRun.dag_id == DagModel.dag_id,
+                DagRun.execution_date == recent_runs_subquery.c.execution_date,
+            ),
+        )
+        .where(recent_runs_subquery.c.rank <= dag_runs_limit.value)
+        .group_by(
+            DagModel.dag_id,
+            recent_runs_subquery.c.execution_date,
+            DagRun.start_date,
+            DagRun.end_date,
+            DagRun.state,
+            DagRun.data_interval_start,
+            DagRun.data_interval_end,
+        )
+    )
+    dags_with_recent_dag_runs_select_filter, total_entries = paginated_select(
+        dags_with_recent_dag_runs_select,
+        [only_active, paused, dag_id_pattern, dag_display_name_pattern, tags, owners, last_dag_run_state],
+        None,
+        offset,
+        limit,
+    )
+    dags_with_recent_dag_runs = session.execute(dags_with_recent_dag_runs_select_filter).all()
+    # aggregate rows by dag_id
+    dag_runs_by_dag_id: dict[str, list] = {}
+    for row in dags_with_recent_dag_runs:
+        dag_id = row.dag_id
+        if dag_id not in dag_runs_by_dag_id:
+            dag_runs_by_dag_id[dag_id] = []
+        dag_runs_by_dag_id[dag_id].append(row)
+
+    return RecentDAGRunsCollectionResponse(
+        total_dag_runs=total_entries,
+        total_dag_ids=len(dag_runs_by_dag_id),
+        recent_dag_runs=[
+            RecentDAGRunsResponse(
+                dag_id=dag_id,
+                dag_runs=[
+                    RecentDAGRun(
+                        start_date=dag_run.start_date,
```

Comment:
I think we should try to keep public and private serializers the same unless we specifically need a new representation of the object. (This will avoid drifts and make things easier to integrate to the public API).

Also in the response I think DagRuns should be grouped per Dag. For instance, you can subclass the public `DAGResponse` and add a `latest_dag_runs` field to it. And then reuse the `DAGCollectionResponse` from there. And same you can re-use the public `DagRunResponse` to serialize the DagRun.

Basically the UI endpoint returns exactly the same but with an additional `latest_dag_runs` field for each DAGs.

Reply:
Sure, I will refactor this UI endpoint using the public serializers.